aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java

```diff
@@ -66,10 +70,14 @@ class S3InputStream extends SeekableInputStream implements RangeReadable {
   private int skipSize = 1024 * 1024;
   private RetryPolicy<Object> retryPolicy =
       RetryPolicy.builder()
-          .handle(
-              ImmutableList.of(
-                  SSLException.class, SocketTimeoutException.class, SocketException.class))
-          .onFailure(failure -> openStream(true))
+          .handle(RETRYABLE_EXCEPTIONS)
+          .onRetry(
+              e -> {
+                LOG.warn(
+                    "Retrying read from S3, reopening stream (attempt {})", e.getAttemptCount());
+                resetForRetry();
+              })
+          .onFailure(e -> LOG.error("Failed to read from S3 after retry policy", e.getException()))
```

Comment:
"Failed to read from S3 input stream after exhausting all retries"

Reply:
Done.